Add shelljs (v0.7.0) declarations

freebroccolo opened the pull request:
This PR adds flow declarations for `shelljs`. It covers the complete API but in some cases I have restricted the interfaces for certain functions to the part that can be reasonably typed. However, this should not limit functionality.

I've added tests following some of the others as examples but I had some trouble getting the `flow-typed` CLI tool to run them.

jeffmo commented on definitions/npm/shelljs_v0.7.x/flow_>=v0.28.x/shelljs_v0.7.x.js:
```diff
@@ -0,0 +1,381 @@
+  & ((rest: void) => $npm$shelljs$String)
+  ;
+
+declare module 'shelljs' {
```
Since it looks like this module only defined named exports, I would recommend 2 tweaks here:

1. You can switch all the `declare`s inside your `declare module` body here to be `declare export` 
2. You can then move all of the global `$npm$shelljs$*` stuff inside the `declare module` body and, because you're not using `declare export`, they will be hidden from the exports of other modules.

There was a practice of using namespaced globals in other libdefs before because there wasn't a good option of hiding `declare`s inside a `declare module` before -- but that was fixed for ES module `declare module`s in v0.28!

freebroccolo replied:
Interesting. I just experimented with `declare export` and I think I see what you mean. I should be able to clean some of this up that way.

Regarding namespaced globals, the main reason I have been using them in my libdefs so far is so that I can refer to types defined in libdefs for some module `m0` in libdefs for another module `m1`.

For instance, I have another small library built on `shelljs` that provides a promise based API for some parts like `exec`. I use flow for the code and babel + rollup to transform it to normal javascript on prepublish. Along with the package, I provide a libdef that refers to the types defined in this one.  In order to make the libdef for my library usable, it seemed like I needed to use namespaced globals for the `shelljs` libdef.

Is there another way to preserve this type information across module/package boundaries? I've been keeping an eye on https://github.com/flowtype/flow-typed/issues/16 since it seems relevant.

Just to be clear, here's a simplified example of the scenario I'm describing:

```javascript
// .flowconfig
[libs]
libdef-m0.js
libdef-m1.js

// libdef-m0.js
declare module "m0" {
  declare interface I0 {
    p0: T0;
  };
}

// libdef-m1.js
declare module "m1" {
  declare interface I1 {
    // How do we provide the type for p0? We can `import type { … }` in this
    // file (libdef-m1.js) but flow doesn't seem to preserve that information
    // when we use this as a lib from another file (see below).
    p0: m0.I0;
    p1: string;
  };
}

// index.js
// @flow
import type { I1 } from "m1";

declare var foo: I1;
// foo.p0 <- flow reports no type
// foo.p1 <- flow reports type string
```

jeffmo replied:
> the main reason I have been using them in my libdefs so far is so that I can refer to types defined in libdefs for some module m0 in libdefs for another module m1.

Ah ok, yea that's definitely useful. If that's your purpose then fine by me until #16 is resolved

freebroccolo replied:
Okay cool. I'll be happy to update this to remove the namespace stuff once there's a resolution.

Aside from that, there are a couple things I'm cleaning up a bit in this PR but once that's done it should be good to go from my side.

jeffmo commented:
I love how complete and extensively tested this libdef is! Awesome :D

freebroccolo commented:
Okay, I cleaned up some of the exports and moved most of the stuff that's probably not needed in other libdefs into the module declaration. Think it's good to go on my end.

jeffmo commented:
![woo!](https://cdn1.communications.syr.edu/content/bb626b6e732642a38bfba5486c2c8965/13270577.gif)
